### PR TITLE
Enhance/meta 3919 - Create refresh classification task(Master PR)

### PR DESCRIPTION
**Pull request by sumandas0**
## To reduce the no of task created we have introduced new type of task which will ensure classification is only propagated to reachable assets and deleted from unreachable assets

> We have created a new type of task `CLASSIFICATION_REFRESH_PROPAGATION` which will be scheduled in case of any propagation removal is needed. And if this task is already scheduled for one vertex id and it is in the pending state then we will not schedule any more refresh propagation tasks.

## Type of change
- [x] New feature (adds functionality)

### Development

- [x] Application changes have been tested thoroughly

### CODE REVIEW

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v1/DeleteHandlerV1.java [hunk 1]
```diff
@@ -1351,23 +1347,66 @@ public void createAndQueueTask(String taskType, AtlasEdge edge) {
...
 
-            AtlasVertex referenceVertex = GraphHelper.getPropagatingVertex(edge);
-            if(referenceVertex == null) {
-                return;
+            boolean removePropagationOnEntityDelete = GraphHelper.getRemovePropagations(currentClassificationVertex);
+
+            if (!(isTermEntityEdge || removePropagationOnEntityDelete)) {
+                LOG.debug("This edge is not term edge or remove propagation isn't enabled");
+                continue;
             }
```
Comment (nikhilbonte21) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v1/DeleteHandlerV1.java:1369]: Please move this code block before the previous if statement
Why? 
imagine f (!(isTermEntityEdge || removePropagationOnEntityDelete))  is already resulting into true then we will save on making ES query

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v1/DeleteHandlerV1.java [hunk 2]
```diff
@@ -1351,23 +1347,66 @@ public void createAndQueueTask(String taskType, AtlasEdge edge) {
...
+        List<AtlasTask> queuedTasks  = RequestContext.get().getQueuedTasks();
+        for(AtlasTask queuedTask : queuedTasks) {
+            if(queuedTask.getClassificationId().equals(classificationId)) {
+                String queuedTaskType = queuedTask.getType();
+                if(queuedTaskType.equals(CLASSIFICATION_REFRESH_PROPAGATION)
+                        || queuedTaskType.equals(CLASSIFICATION_PROPAGATION_DELETE)) {
+                    return true;
+                }
+            }
+        }
```
Comment (nikhilbonte21) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v1/DeleteHandlerV1.java:1404]: Move this up as well before hitting ES check

File: repository/src/main/java/org/apache/atlas/repository/graph/GraphHelper.java [hunk 3]
```diff
@@ -390,6 +390,30 @@ public static AtlasEdge getClassificationEdge(AtlasVertex entityVertex, AtlasVer
...
+                .direction(AtlasEdgeDirection.OUT)
+                .label(CLASSIFICATION_LABEL)
+                .has(CLASSIFICATION_EDGE_NAME_PROPERTY_KEY, getTypeName(classificationVertex))
+                .vertices().iterator();
+
+        if (vertices != null) {
+            while (vertices.hasNext()) {
+                AtlasVertex vertex = vertices.next();
+                if (vertex != null) {
+                    if (vertex.getIdForDisplay().equals(classificationVertex.getIdForDisplay())) {
```
[the remark this answers is not in the document]
Comment (sumandas0) [reply repository/src/main/java/org/apache/atlas/repository/graph/GraphHelper.java]: Compiler handles that, still we can improve this. Modified the code

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v1/DeleteHandlerV1.java [hunk 4]
```diff
@@ -424,6 +428,7 @@ private void addTagPropagation(AtlasVertex fromVertex, AtlasVertex toVertex, Atl
 
         if (taskManagement != null && DEFERRED_ACTION_ENABLED) {
             for (AtlasVertex classificationVertex : classificationVertices) {
+                //If any Refresh classification or delete classification task pending don't schedule
```
[the remark this answers is not in the document]
Comment (sumandas0) [reply repository/src/main/java/org/apache/atlas/repository/store/graph/v1/DeleteHandlerV1.java]: Sure

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v1/DeleteHandlerV1.java (hunk 1 above)
Comment (sumandas0) [reply repository/src/main/java/org/apache/atlas/repository/store/graph/v1/DeleteHandlerV1.java]: Yes, you are right. Changed this

**Review by mehtaanshul — APPROVED** [review]

**Review by nikhilbonte21 — APPROVED** [review]

**Pull request merged**